--- a/src/ARMJIT.cpp
+++ b/src/ARMJIT.cpp
@@ -940,7 +940,7 @@
 
     AddressRange* region = CodeMemRegions[localAddr >> 27];
     AddressRange* range = &region[(localAddr & 0x7FFFFFF) / 512];
-    u32 mask = 1 << ((localAddr & 0x1FF) / 16);
+    u32 writeMask = 1 << ((localAddr & 0x1FF) / 16);
 
     range->Code = 0;
     for (int i = 0; i < range->Blocks.Length;)
@@ -948,21 +948,21 @@
         JitBlock* block = range->Blocks[i];
 
         bool invalidated = false;
-        u32 mask = 0;
+        u32 blockMask = 0;
         for (int j = 0; j < block->NumAddresses; j++)
         {
             if (block->AddressRanges()[j] == (localAddr & ~0x1FF))
             {
-                mask = block->AddressMasks()[j];
-                invalidated = block->AddressMasks()[j] & mask;
-                assert(mask);
+                blockMask = block->AddressMasks()[j];
+                invalidated = (blockMask & writeMask) != 0;
+                assert(blockMask);
                 break;
             }
         }
-        assert(mask);
+        assert(blockMask);
         if (!invalidated)
         {
-            range->Code |= mask;
+            range->Code |= blockMask;
             i++;
             continue;
         }
--- a/src/ARMJIT_A64/ARMJIT_LoadStore.cpp
+++ b/src/ARMJIT_A64/ARMJIT_LoadStore.cpp
@@ -211,10 +211,18 @@
         else
         {
             LDRGeneric(size, flags & memop_SignExtend, rdMapped, size > 8 ? X1 : X0, RMemBase);
-            if (size == 32 && !addrIsStatic)
-            {
-                UBFIZ(W0, W0, 3, 2);
-                RORV(rdMapped, rdMapped, W0);
+            if (size == 32)
+            {
+                if (addrIsStatic)
+                {
+                    if (staticAddress & 0x3)
+                        ROR(rdMapped, rdMapped, (staticAddress & 0x3) << 3);
+                }
+                else
+                {
+                    UBFIZ(W0, W0, 3, 2);
+                    RORV(rdMapped, rdMapped, W0);
+                }
             }
         }
 
